### PR TITLE
Create listExpiredISAs and listExpiredSubscriptions

### PR DESCRIPTION
**Pull request by Dos-Ph**
#423 
Next:
1. set scheduler to call these functions
2. add Garbage collector detail to 'build/README.md' file

### CODE REVIEW

File: pkg/rid/store/cockroach/identification_service_area.go [hunk 1]
```diff
@@ -222,3 +222,25 @@ func (c *isaRepo) SearchISAs(ctx context.Context, cells s2.CellUnion, earliest *
...
+		ends_at > $1
+	AND
+		writer = $2`, isaFields)
+	)
+
+	if len(writer) == 0 {
+		return nil, stacktrace.NewError("Writer is required")
+	}
+
+	return c.process(ctx, isasInCellsQuery, expiredTime, writer)
```
Comment (charlie-pisuraj) [inline pkg/rid/store/cockroach/identification_service_area.go:245]: ok I think I might see a problem here, if you passed in `expiredTime` we run the risk of using in consistent time from different servers. I understand normally we all use unix epoch but we probably don't want to risk it. We should probably just query it with the SQL statement for getting the database time like `CURRENT_TIME` and do offset with it.

**Review by charlie-pisuraj — CHANGES_REQUESTED** [review]

**Comment by charlie-pisuraj** [comment]
This looks good to me, maybe make the 30 Minute a configurable value? but other than that I'm fine with it

@vosst can you take a look?

**Review by charlie-pisuraj — APPROVED** [review]

File: pkg/rid/store/cockroach/identification_service_area.go [hunk 2]
```diff
@@ -222,3 +222,25 @@ func (c *isaRepo) SearchISAs(ctx context.Context, cells s2.CellUnion, earliest *
 
 	return c.process(ctx, isasInCellsQuery, earliest, latest, pq.Int64Array(cids))
 }
+
+// ListExpiredISAs lists all expired ISAs based and writer.
```
Comment (BenjaminPelletier) [inline pkg/rid/store/cockroach/identification_service_area.go:226]: "...all expired ISAs based *on* writer."

File: pkg/rid/store/cockroach/identification_service_area.go [hunk 3]
```diff
@@ -222,3 +222,25 @@ func (c *isaRepo) SearchISAs(ctx context.Context, cells s2.CellUnion, earliest *
...
+	FROM
+		identification_service_areas
+	WHERE
+		ends_at + INTERVAL '%d' MINUTE <= CURRENT_TIMESTAMP
+	AND
+		writer = $1`, isaFields, expiredDurationInMin)
+	)
+
+	if len(writer) == 0 {
+		return nil, stacktrace.NewError("Writer is required")
```
Comment (BenjaminPelletier) [inline pkg/rid/store/cockroach/identification_service_area.go:242]: How will ISAs with no writer be removed if we can't specify an empty string here?

File: pkg/rid/store/cockroach/subscriptions.go [hunk 4]
```diff
@@ -301,3 +301,25 @@ func (c *subscriptionRepo) SearchSubscriptionsByOwner(ctx context.Context, cells
 
 	return c.process(ctx, query, pq.Int64Array(cids), owner, c.clock.Now())
 }
+
+// ListExpiredSubscriptions lists all expired Subscriptions based and writer.
```
Comment (BenjaminPelletier) [inline pkg/rid/store/cockroach/subscriptions.go:305]: "...based *on* writer."

**Review by BenjaminPelletier — COMMENTED** [review]
This looks good to me with the exception that I think we need to have a way to deal with empty-writer records.  Right now, we can't even query for them, so we definitely won't be able to delete them.  I think an easy fix would be to allow the writer argument to be empty when listing expired resources, and retrieving no-writer entries in that case.  That will allow us to target them for deletion under appropriate circumstances.

File: pkg/rid/store/cockroach/identification_service_area.go (hunk 3 above)
Comment (Dos-Ph) [reply pkg/rid/store/cockroach/identification_service_area.go]: removed condition and add ListExpiredSubscriptionsWithNullWriter function to remove expired record with writer = NULL (writer can be NULL when db is migrated from v3.0 to v3.1)

**Comment by BenjaminPelletier** [comment]
Is there a case where we would want to retrieve entities with writer='' separately from entities with writer=null?  If not, I suggest we retrieve any entities with writer='' or writer=null when a blank writer is passed to ListExpired*.  This simplifies the API by letting us eliminate the separate ListExpired*WithNullWriter methods.

File: pkg/rid/store/cockroach/identification_service_area.go [hunk 5]
```diff
@@ -222,3 +222,26 @@ func (c *isaRepo) SearchISAs(ctx context.Context, cells s2.CellUnion, earliest *
 
 	return c.process(ctx, isasInCellsQuery, earliest, latest, pq.Int64Array(cids))
 }
+
+// ListExpiredISAs lists all expired ISAs based on writer.
+// Records expire if current time is <expiredDurationInMin> minutes more than records' endTime.
+func (c *isaRepo) ListExpiredISAs(ctx context.Context, writer *string) ([]*ridmodels.IdentificationServiceArea, error) {
+	writerQuery := "NULL"
+	if writer != nil {
```
Comment (BenjaminPelletier) [inline pkg/rid/store/cockroach/identification_service_area.go:230]: With this implementation, if we want to garbage-collect everything without a writer in a robust way, we'll need to call ListExpiredISAs(ctx, "") and ListExpiredISAs(ctx, nil) -- is there ever a case in which we would want to make those calls separately, or distinguish between entries with "" writer and null writer?  If not, let's accept a standard string argument and query for both "" and null when the empty string is supplied.

**Review by BenjaminPelletier — APPROVED** [review]

**Pull request merged**